### Baudouin M. Ramazani
### Quantum Physics - Project I
### Prof: Dr. Steve Spicklemire

##                        Representing Complex Numbers Visually with VPython

One of the primary learning goals of this course is for you to develop an intuitive understanding of what it means to do quantum mechanics and how the machinery works. I have found that it helps many students to have a concrete visual image or representation of what's going on. Since quantum mechanical probability amplitudes are actually complex numbers, it's necessary to have a visual representation of a complex number.

### A Phasor getting multiplied by a small pure phase.
Let's say we start with a complex number like $z=2e^{i\pi/4}$. What happens to that phasor if we repeatedly multiply by a complex number with magnitude of one, but with a small non-zero phase angle $c=e^{i\delta}$? Let's find out! First let's get a good representation of our original complex number. Type in the following code and run it. Make sure you understand each line. If you have any questions, please ask. When the program runs you should see a 3D representation of our starting complex number. Does it look reasonable? Is it pointing in the right direction?

In [41]:
from vpython import *
from numpy import *
import cmath
from vpython import rate


In [42]:
def SetArrowFromCN(cn, a):
    """
    SetArrowWithCN takes a complex number 'cn' and an arrow object 'a'.
    It sets the 'x' and 'y' components of the arrow's axis to the real
    and imaginary parts of the given complex number.
    """
    a.axis.x = cn.real
    a.axis.y = cn.imag


In [45]:
N = 628
PHASE = 0.01
i=1j                            # Let's use 'i' for "imaginary"
a = arrow(color=color.red)      # red is a nice color
z = 2*exp(i*pi/4.0)
count=0
c = 1*exp(i*PHASE)
phasors = []

for i in range(N):
    z = z*c
    phasors.append(z)
    print("Phasor #",i+1,": ",z," => Magnitude: ",abs(z)," => Angle: ",cmath.phase(complex(z.real, z.imag)))


Phasor # 1 :  (1.400000952361581+1.4282847522068791j)  => Magnitude:  2.0  => Angle:  0.7953981633974483
Phasor # 2 :  (1.3856483434214943+1.4422131147556758j)  => Magnitude:  1.9999999999999998  => Angle:  0.8053981633974483
Phasor # 3 :  (1.3711571708017687+1.4559972571948372j)  => Magnitude:  1.9999999999999998  => Angle:  0.8153981633974483
Phasor # 4 :  (1.35652888360759+1.4696358011216062j)  => Magnitude:  1.9999999999999998  => Angle:  0.8253981633974483
Phasor # 5 :  (1.3417649446554873+1.4831273826929554j)  => Magnitude:  1.9999999999999998  => Angle:  0.8353981633974484
Phasor # 6 :  (1.326866830327053+1.4964706527619707j)  => Magnitude:  1.9999999999999998  => Angle:  0.8453981633974484
Phasor # 7 :  (1.3118360304213046+1.5096642770127644j)  => Magnitude:  1.9999999999999996  => Angle:  0.8553981633974485
Phasor # 8 :  (1.296674048005707+1.5227069360939063j)  => Magnitude:  1.9999999999999996  => Angle:  0.8653981633974485
Phasor # 9 :  (1.281382399265867+1.5355973257503568j

Phasor # 227 :  (-1.992575108498469+0.17217559929409507j)  => Magnitude:  1.9999999999999916  => Angle:  3.0553981633974483
Phasor # 228 :  (-1.9941972078704322+0.1522415715950762j)  => Magnitude:  1.9999999999999916  => Angle:  3.0653981633974485
Phasor # 229 :  (-1.9956198891834338+0.1322923198657654j)  => Magnitude:  1.9999999999999916  => Angle:  3.0753981633974483
Phasor # 230 :  (-1.9968430101705281+0.11232983901471125j)  => Magnitude:  1.9999999999999916  => Angle:  3.0853981633974485
Phasor # 231 :  (-1.9978664485206357+0.09235612527336351j)  => Magnitude:  1.9999999999999916  => Angle:  3.0953981633974483
Phasor # 232 :  (-1.9986901018907743+0.07237317599645161j)  => Magnitude:  1.9999999999999918  => Angle:  3.1053981633974486
Phasor # 233 :  (-1.9993138879162935+0.05238298946225084j)  => Magnitude:  1.999999999999992  => Angle:  3.1153981633974484
Phasor # 234 :  (-1.9997377442191102+0.03238756467275619j)  => Magnitude:  1.9999999999999918  => Angle:  3.1253981633974486
Phas

Phasor # 414 :  (0.4228040511969316-1.9547983871211385j)  => Magnitude:  1.9999999999999853  => Angle:  -1.3577871437821378
Phasor # 415 :  (0.4423305692436487-1.950472677971299j)  => Magnitude:  1.9999999999999853  => Angle:  -1.3477871437821378
Phasor # 416 :  (0.46181285460204907-1.9459519231790512j)  => Magnitude:  1.9999999999999853  => Angle:  -1.3377871437821378
Phasor # 417 :  (0.48124895905983195-1.9412365748161067j)  => Magnitude:  1.9999999999999853  => Angle:  -1.3277871437821378
Phasor # 418 :  (0.5006369390227483-1.9363271044133723j)  => Magnitude:  1.9999999999999853  => Angle:  -1.317787143782138
Phasor # 419 :  (0.5199748557089585-1.931224002913797j)  => Magnitude:  1.9999999999999853  => Angle:  -1.3077871437821378
Phasor # 420 :  (0.5392607753429087-1.9259277806232784j)  => Magnitude:  1.9999999999999853  => Angle:  -1.2977871437821378
Phasor # 421 :  (0.5584927693487072-1.9204389671596318j)  => Magnitude:  1.9999999999999853  => Angle:  -1.2877871437821378
Phasor # 

In [46]:
canvas()
for i in range(len(phasors)):
    rate(100)
    SetArrowFromCN(phasors[i],a)


<IPython.core.display.Javascript object>

### Questions
Please answer these questions at the end of your report.

1) What happens to the magnitude of the complex number 'cn' during the execution of your loop?

##### => The magnitude of the phasor does not show a change on the canvas. However, the numbers obtained show a very small change less than 10^-15.

2) What happens to the phase of the complex number 'cn' during the execution of your loop?

#### => The phase is basicaly what changes the position of the arrow. Therefore,  by simply looking, we can tell that the angle of the phasor changes since it turns counterclockwise.  On top of that, the numbers show that only the second decimal digit of the phase changes.

3) How many iterations of your loop does it take for the phasor representation to complete a full cycle? Can you explain why this is a reasonable amount of time?

#### It takes exactly 360 degrees or $2\pi$ radians to complete a full counterclockwise turn. Therefore, we need about 314x2 = 628 iterations given a 0.01 phase change.

4) What is the cartesian form of the number c? Demonstrate analytically that multiplying by this number has the effect you observed on both the magnitude and the angle of the phasor?


$$ c = e^{\rm i\delta} $$

###### We chose $$\delta = 0.01 $$ and  the amplitude of the phasor as 1

###### This means 

$$ c(e^{\rm i\delta}) = c(cos(0.01) + isin(0.01)) =  c(0.99995 +  0.00999i)$$


$$ Z' = (1 + 0.01i)(1.40 + 1.42i) = (1.40  - 0.01456) + i(1.38544 + 0.0137) =  1.38544 + i1.4337 $$

###### Let's apply pythagore's theorem:

$$ |x| = \sqrt{\rm (1.355)^2 + (1.4697i)^2} = 1.99 $$
###### As computer by python, the result is close to 2
###### Then, we also get the same angle
$$ {\theta} = Arctan({\frac{1.4697}{1.355}}) = 0.80 rad$$

#### =====================================================
#### Please, restart the kernel if the arrow does not update its position.
#### =====================================================